core-patch/src/test/scala/eu/ostrzyciel/jelly/core/patch/helpers/PatchTestCases.scala

```diff
@@ -0,0 +1,311 @@
+package eu.ostrzyciel.jelly.core.patch.helpers
+
+import eu.ostrzyciel.jelly.core.helpers.Mrl.*
+import eu.ostrzyciel.jelly.core.proto.v1.*
+import eu.ostrzyciel.jelly.core.proto.v1.patch.*
+
+import scala.annotation.experimental
+
+@experimental
+object PatchTestCases:
+  import Mpl.*
+  import RdfPatchRow as R
+
+  val testCases: Seq[(String, PatchTestCase, PatchStatementType.Recognized)] = Seq(
+    ("a triple patch", Triples1, PatchStatementType.TRIPLES),
+    ("a triple patch with namespace declarations", Triples2NsDecl, PatchStatementType.TRIPLES),
+    ("a quad patch", Quads1, PatchStatementType.QUADS),
+    ("nonsensical transaction commands", MalformedTransactions, PatchStatementType.TRIPLES),
+  )
+
+  trait PatchTestCase:
+    def mrl: Seq[PatchStatement]
+    def encoded(opt: RdfPatchOptions): Seq[RdfPatchRow]
+    final def encodedFull(opt: RdfPatchOptions, groupByN: Int): Seq[RdfPatchFrame] =
+      encoded(opt)
+        .grouped(groupByN)
+        .map(rows => RdfPatchFrame(rows))
+        .toSeq
+      
+  object Triples1 extends PatchTestCase:
+    val mrl = Seq(
+      Header("key", Iri("https://test.org/test/subject")),
+      TxStart,
+      Add(Triple(
+        Iri("https://test.org/test/subject"),
+        Iri("https://test.org/test/predicate"),
+        Iri("https://test.org/ns2/object"),
+      )),
+      Add(Triple(
+        Iri("https://test.org/test/subject"),
+        Iri("https://test.org/test/predicate"),
+        DtLiteral("123", Datatype("https://test.org/xsd/integer")),
+      )),
+      Delete(Triple(
+        Iri("https://test.org/test/subject"),
+        Iri("https://test.org/test/predicate"),
+        Iri("https://test.org/ns2/object"),
+      )),
+      TxCommit,
+      TxStart,
+      Add(Triple(
+        Iri("https://test.org/test/subject"),
+        Iri("https://test.org/test/predicate"),
+        TripleNode(Triple(Iri("https://test.org/test/subject"), Iri("b"), Iri("c"))),
+      )),
+      TxAbort,
+      TxStart,
+      Add(Triple(
+        Iri("https://test.org/test/predicate"),
+        Iri("https://test.org/test/subject"),
+        TripleNode(Triple(Iri("https://test.org/test/subject"), Iri("b"), Iri("c"))),
+      )),
+      Delete(Triple(
+        Iri("https://test.org/test/predicate"),
+        Iri("https://test.org/test/subject"),
+        TripleNode(Triple(Iri("https://test.org/test/subject"), Iri("b"), Iri("c"))),
+      )),
+      TxCommit,
+    )
+
+    def encoded(opt: RdfPatchOptions): Seq[RdfPatchRow] = Seq(
+      R.ofOptions(opt),
+      R.ofPrefix(RdfPrefixEntry(0, "https://test.org/test/")),
+      R.ofName(RdfNameEntry(0, "subject")),
+      R.ofHeader(RdfPatchHeader("key", RdfIri(1, 0))),
+      R.ofTransactionStart,
+      R.ofName(RdfNameEntry(0, "predicate")),
+      R.ofPrefix(RdfPrefixEntry(0, "https://test.org/ns2/")),
+      R.ofName(RdfNameEntry(0, "object")),
+      R.ofStatementAdd(RdfQuad(
```

Comment:
I would move some of this sequence building logic to dedicated helpers, maybe at least the triple and quad construction logic - you could have a helper class keeping track of the statements it generated and either returning one of the existing statements to you (so that you could test deleting it) or generate a new statement by having a new subject, predicate or object. Otherwise the testing sequences may be a bit hard to read/maintain, but I guess if the patterns do not overlap enough you may now want to do it

Reply:
I'm not convinced that helpers would be much more maintainable. There are lots of very, very different test cases here. Writing helpers for that would be an utter headache.